types/vote.go

```diff
@@ -61,6 +61,53 @@ type Vote struct {
 	ExtensionSignature []byte                `json:"extension_signature"`
 }
 
+func voteFromProto(pv *tmproto.Vote) (*Vote, error) {
+	if pv == nil {
+		return nil, errors.New("nil vote")
+	}
+
+	blockID, err := BlockIDFromProto(&pv.BlockID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Vote{
+		Type:               pv.Type,
+		Height:             pv.Height,
+		Round:              pv.Round,
+		BlockID:            *blockID,
+		Timestamp:          pv.Timestamp,
+		ValidatorAddress:   pv.ValidatorAddress,
+		ValidatorIndex:     pv.ValidatorIndex,
+		Signature:          pv.Signature,
+		Extension:          pv.Extension,
+		ExtensionSignature: pv.ExtensionSignature,
+	}, nil
+}
+
+// VoteFromProto attempts to convert the given serialization (Protobuf) type to
+// our Vote domain type, performing basic validation with ValidateBasic. Note
+// that this performs no vote extension validation - to additionally validate
+// vote extensions, use VoteWithExtensionFromProto instead.
+func VoteFromProto(pv *tmproto.Vote) (*Vote, error) {
```

Comment:
I find the names of these two slightly misleading: "VoteFromProto" suggests I'm getting the entire vote, implying all its parts (including the extension). Reading "VoteWithExtensionFromProto" feels like "this one HAS the extension, the other doesn't". 

But both _return_ the extension—it's just this one doesn't verify it while the other does.

If this were Java, we'd just say `ExtractAndVerifyVoteWithoutExtensionFromProto` and `ExtractAndVerifyVoteWithExtensionFromProto`. But here, I wonder if maybe we should consider separating "extract" and "verify", i.e., just let the caller call `ValidateBasic` or `ValidateWithExtension` as they prefer. I don't think we're adding much value by hiding that call, when they already have to know which they want.

Reply:
Yes, it does make sense to rather separate the logic for extracting and validating votes. I think it's usually better to have each function just do one thing. I'll separate them :+1: